### DISCUSSION TITLE
OTGW - Home Assistant Integration

### DISCUSSION THREAD

**Discussion by repressed-geek** [rvdbreemen/OTGW-firmware#44] [Q&A] [answered]
Hi Robert,

Thank you for all of your work, you've obviously spent a lot of time working on this!

This is likely to be a terribly noob question....

I have a Nodoshop OTGW with NodeMCU. I downloaded your firmware to the NodeMCU, connected it to my network and plugged the board into the OTGW. I can access the GW via the web portal and get the Home Screen. I have also updated the PIC firmware to v5.1

I have tried to add the OTGW as an integration on HA but every time I try to add the integration it tries for ~10 secs and then states 'Failed to Connect'. I've tried restarting both but can get no further. Have I missed something? Any assistance would be very gratefully received.

Kind regards,

Tom

**Comment by rvdbreemen** [comment, marked as answer]
Hi Tom

Mqtt is reconnected every 10minutes. I’d it cannot connect after 3 tries. 

Just a few questions:
1. Do you have MQTT running with your Home Assistant setup?
2. Did you setup the MQTT broker on the settingpage?
3. Have you connected to your broker with an MQTT client (for example MQTT Explorer)?
4. Can you capture some debug logs for me?

Let me know what you find,
Robert

**Reply by repressed-geek** [reply]
Robert,

Thank you for such a fast response!!

1. I was borderline clueless about MQTT, I have spent my morning finding out about it. I did have the MQTT Broker integration running but it wasn't properly configured. I have since installed the Mosquitto broker and reinstalled the MQTT Integration. Following some initial teething it can now see the OTGW.

2. Due to the above no, but it now is.

3. Just installed it and connected to the Mosquitto broker, I have attached a screenshot.
<img width="1027" alt="Screenshot 2021-03-27 at 14 03 20" src="https://user-images.githubusercontent.com/81447397/112723221-61010f80-8f05-11eb-9055-92ca57cd8930.png">

4. Broker log:

Mosquitto Broker

[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[11:50:14] INFO: Setup mosquitto configuration
[11:50:14] WARNING: SSL not enabled - No valid certs found!
[11:50:14] INFO: No local user available
[11:50:14] INFO: Initialize system configuration.
[11:50:14] INFO: Initialize Home Assistant Add-on services
[11:50:14] INFO: Initialize Home Assistant discovery
[11:50:14] INFO: Start Mosquitto daemon
1616845814: mosquitto version 1.6.3 starting
1616845814: Config loaded from /etc/mosquitto.conf.
1616845814: Loading plugin: /usr/share/mosquitto/auth-plug.so
1616845814:  ├── Username/password checking enabled.
1616845814: |-- *** auth-plug: startup
1616845814:  ├── TLS-PSK checking enabled.
1616845814:  └── Extended authentication not enabled.
1616845814: Opening ipv4 listen socket on port 1883.
1616845814: Opening ipv6 listen socket on port 1883.
1616845814: Opening websockets listen socket on port 1884.
1616845814: Warning: Mosquitto should not be run as root/administrator.
1616845823: New connection from 172.30.33.0 on port 1883.
[11:50:24] ERROR: [ERROR] Auth error with mqtt
1616845824: Socket error on client <unknown>, disconnecting.
1616845843: New connection from 172.30.32.1 on port 1883.
[11:50:43] INFO: [INFO] found homeassistant on local database
1616845843: New client connected from 172.30.32.1 as auto-E7D5402B-B85F-0578-5D80-2FCE11D9E09C (p2, c1, k60, u'homeassistant').
1616845843: Client auto-E7D5402B-B85F-0578-5D80-2FCE11D9E09C disconnected.
1616845843: New connection from 172.30.32.1 on port 1883.
1616845843: New client connected from 172.30.32.1 as 1MLQviv44YaHLOofQuPlC8 (p2, c1, k60, u'homeassistant').
1616846064: New connection from 192.168.1.13 on port 1883.
1616846064: Socket error on client <unknown>, disconnecting.
1616846071: New connection from 192.168.1.13 on port 1883.
1616846071: Socket error on client <unknown>, disconnecting.
1616846078: New connection from 192.168.1.13 on port 1883.
1616846078: Socket error on client <unknown>, disconnecting.
1616846086: New connection from 192.168.1.13 on port 1883.
1616846086: Socket error on client <unknown>, disconnecting.
1616846093: New connection from 192.168.1.13 on port 1883.
1616846093: Socket error on client <unknown>, disconnecting.
1616846278: New connection from 192.168.1.13 on port 1883.
[11:57:59] INFO: [INFO] found mqttuser on Home Assistant
1616846279: New client connected from 192.168.1.13 as OTGWF4:CF:A2:F4:BF:DF (p2, c1, k15, u'mqttuser').
1616847615: Saving in-memory database to /data/mosquitto.db.
1616849416: Saving in-memory database to /data/mosquitto.db.
1616851217: Saving in-memory database to /data/mosquitto.db.
1616853018: Saving in-memory database to /data/mosquitto.db.
1616853263: Client OTGWF4:CF:A2:F4:BF:DF has exceeded timeout, disconnecting.

**Reply by rvdbreemen** [reply]
Well, that looks like you are now receiving messages on you MQTT broker. Did you configure Home Assistant to point to your broker too? If so, you should see the OTGW device as a discovered device.

**Reply by repressed-geek** [reply]
The OTGW comes up as a device on the MQTT Integration. The individual OpenTherm elements show up as Binary Sensors and Sensors on the overview.

<img width="385" alt="Screenshot 2021-03-27 at 14 24 09" src="https://user-images.githubusercontent.com/81447397/112723795-3c5a6700-8f08-11eb-87b5-c7c0af56fa90.png">
<img width="376" alt="Screenshot 2021-03-27 at 14 24 22" src="https://user-images.githubusercontent.com/81447397/112723797-3d8b9400-8f08-11eb-9720-6dfd31c1795d.png">

**Reply by repressed-geek** [reply]
But the OTGW does not show up as a discovered device.

**Reply by rvdbreemen** [reply]
Could you check this for me:
1. Goto settings in Home Assistant 
2. Goto integrations
3. Goto MQTT integration

At this point you should see one device (at least). If you only have OTGW you will see 1 device and 65 entity’s. 

Let me know what you see please. 
Robert

**Reply by repressed-geek** [reply]
Yes I have 1 device and 65 entities.

**Comment by sjorsjuhmaniac** [comment]
@repressed-geek in your opening post it sounds like you installed the OTGW INTEGRATION in HA, is that correct? If yes, please remove it. This only works with the 'traditional' nodeMCU firmware named easyESP

Based on the mqttexplorer screenshot I would say the HA Discovery is successfull (see all topics created under 'homeassistant'). Your HA screenshots also confirm you have the entities.

The problem seems to be that there is not 'otgw data' received, there is only 1 topic under 'OTGW'. there should be many more if the otgw is connected to the boiler.

so some questions:
- is the otgw connected to both the boiler and the thermostat?
- do you see any 'live' data on the webUI/webpage ON the nodeMCU "OTGW-Firmware"?
- please share the config settings on the nodeMCU "OTGW-Firmware"
- when you are 'viewing' with mqtt explorer, do you see ANY data coming in at all? Does the 'x message' after "OTGW" change at all?

**Reply by repressed-geek** [reply]
I have tried to install the OTGW integration, I hadn't been able to. It seems from your comment that I never needed the integration which is likely where I've been going wrong. I had assumed that I had.

I have disconnected the OTGW from the boiler because I don't have the Evohome OpenTherm bridge yet and it overrides the BDR91 currently controlling my boiler (the house was quite chilly). When I did have it connected I could read the OT data from the boiler.

NodeMCU Firmware Version0.8.1+92f10cb (14-03-2021)
OTGW PIC Firmware Version5.1
Compiled on (date/time)Mar 14 2021 14:56:26

**Reply by rvdbreemen** [reply]
Okay, well, if the OTGW is not connected, that explains why you see “unknown” values. What you can do is goto the WebUI (http://otgw.local) and check to see if you see any values there. It will only show values, if any OT message are received by the gateway. Only once you see them in the WebUI, you should start seeing them in Home Assistant too.

Also I would ask you to update to the latest firmware, v0.8.2. It fixes some minor Web UI bugs.       

And

**Reply by sjorsjuhmaniac** [reply]
> I have tried to install the OTGW integration, I hadn't been able to. It seems from your comment that I never needed the integration which is likely where I've been going wrong. I had assumed that I had.

@rvdbreemen, maybe something to specifically add to the wiki/readme?
> 
> I have disconnected the OTGW from the boiler because I don't have the Evohome OpenTherm bridge yet and it overrides the BDR91 currently controlling my boiler (the house was quite chilly). When I did have it connected I could read the OT data from the boiler.
> 

I thought the BDR91's send OT signals? I'm guessing it doesn't if you're waiting for an OT bridge ;)

> NodeMCU Firmware Version0.8.1+92f10cb (14-03-2021)
> OTGW PIC Firmware Version5.1
> Compiled on (date/time)Mar 14 2021 14:56:26

Let us know if this issue persists when you have all the hardware setup as it should.

**Reply by repressed-geek** [reply]
Thank you both for your help. I think that I'm there now!

@rvdbreemen, I have updated the firmware, thank you for all your work on this.

@sjorsjuhmaniac, BDR91s are simple relays. OT bridge required. There's a long debate on the [Evohome Shop](https://theevohomeshopforum.co.uk/viewtopic.php?f=5&t=477&sid=ae4629734b4d4c71abfadc97f1cc3520&start=50) forums as the Viessman boilers don't set a max flow temperature (and it isn't possible for the owner to do this), a user called Bart has written code to use the OTGW to force this (otherwise the boiler forever goes like the clappers and overshoots room temperatures). I wanted to get the OTGW in and working before getting the bridge.

@rvdbreemen is it possible to override command values in the OTGW (and thus to the boiler) through home assistant or would I need to edit the firmware?

**Reply by sjorsjuhmaniac** [reply]
Could you copy/paste or create a gist of that code? i'm curious but it's a closed forum. Where is this 'code' executed?

Please elaborate on what you mean by 'override command values'?

**Reply by rvdbreemen** [reply]
You can write commands to OTGW thru the MQTT command topic or REST API. 

Or you could even make a plug-in in this firmware to execute your logic.

Please share the code from the other forum or even join our Discord server to chat about this, right here: https://discord.gg/gRVUQBKHbp

**Reply by repressed-geek** [reply]
I definitely need to be more careful with my language! The user has edited the firmware to limit the max boiler temperature.

> _I've managed to build the gateway firmware with a small and quick modification so that it treats the #57 command in the same manner as the #25 command. This disables the functionality of changing on the wire a write request of the thermostat of the CH max temperature, for sure this shortcoming can be improved upon but since I don't see evohome sending out this command as a write data command (only a read data is done from evohome to the boiler) the hack suits my purpose quite well.

And it works.

I can now override the #57 command response using SR=57:65,00 to pretend that the maximum boiler temperature is 65,00 degrees. Evohome then scales the control setpoint between some minimum and the fake maximum temperature of 65. Log extract:_

22:04:35.399535   Command: SR=57:65,00
22:04:35.440532   SR: 57:65/0
22:04:37.903679   T80190000   Read-Data    Boiler water temperature: 0.00
22:04:38.195700   BC01942CD   Read-Ack     Boiler water temperature: 66.80
22:04:38.437733   T90014C00   Write-Data   Control setpoint: 76.00
22:04:38.706759   B50014C00   Write-Ack    Control setpoint: 76.00
...
22:04:58.382856   T80190000   Read-Data    Boiler water temperature: 0.00
22:04:58.718843   BC019434D   Read-Ack     Boiler water temperature: 67.30
22:04:58.925945   T90014C00   Write-Data   Control setpoint: 76.00
22:04:59.217079   B50014C00   Write-Ack    Control setpoint: 76.00
...
22:05:02.716139   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:03.039131   BC019434D   Read-Ack     Boiler water temperature: 67.30
22:05:03.262193   T90014C00   Write-Data   Control setpoint: 76.00
22:05:03.545155   B50014C00   Write-Ack    Control setpoint: 76.00
...
22:05:07.576364   T00395000   Read-Data    Max CH water setpoint: 80.00
22:05:07.885404   BC0395000   Read-Ack     Max CH water setpoint: 80.00
22:05:07.905126   AC0394100   Read-Ack     Max CH water setpoint: 65.00
22:05:08.125435   T00394100   Read-Data    Max CH water setpoint: 65.00
22:05:08.385432   BC0395000   Read-Ack     Max CH water setpoint: 80.00
22:05:08.391833   AC0394100   Read-Ack     Max CH water setpoint: 65.00
22:05:08.655487   T00394100   Read-Data    Max CH water setpoint: 65.00
22:05:08.890453   BC0395000   Read-Ack     Max CH water setpoint: 80.00
22:05:08.910505   AC0394100   Read-Ack     Max CH water setpoint: 65.00
...
22:05:10.808619   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:11.066606   BC019434D   Read-Ack     Boiler water temperature: 67.30
22:05:11.352594   T90013E00   Write-Data   Control setpoint: 62.00
22:05:11.562626   B50013E00   Write-Ack    Control setpoint: 62.00
...
22:05:12.990673   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:13.224551   B401943CD   Read-Ack     Boiler water temperature: 67.80
22:05:13.519715   T90013E00   Write-Data   Control setpoint: 62.00
22:05:13.719756   B50013E00   Write-Ack    Control setpoint: 62.00
...
22:05:10.808619   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:11.066606   BC019434D   Read-Ack     Boiler water temperature: 67.30
22:05:11.352594   T90013E00   Write-Data   Control setpoint: 62.00
22:05:11.562626   B50013E00   Write-Ack    Control setpoint: 62.00
...
22:05:12.990673   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:13.224551   B401943CD   Read-Ack     Boiler water temperature: 67.80
22:05:13.519715   T90013E00   Write-Data   Control setpoint: 62.00
22:05:13.719756   B50013E00   Write-Ack    Control setpoint: 62.00
...
22:05:18.391972   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:18.702003   B401943CD   Read-Ack     Boiler water temperature: 67.80
22:05:18.918069   T90013E00   Write-Data   Control setpoint: 62.00
22:05:19.206057   B50013E00   Write-Ack    Control setpoint: 62.00
...
22:05:22.709280   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:23.036279   BC01942CD   Read-Ack     Boiler water temperature: 66.80
22:05:23.253401   T90013E00   Write-Data   Control setpoint: 62.00
22:05:23.527313   B50013E00   Write-Ack    Control setpoint: 62.00
...
22:05:28.122525   T80190000   Read-Data    Boiler water temperature: 0.00
22:05:28.381581   B401940E6   Read-Ack     Boiler water temperature: 64.90
...
...
22:08:09.569913   T00394100   Read-Data    Max CH water setpoint: 65.00
22:08:09.754827   BC0395000   Read-Ack     Max CH water setpoint: 80.00
22:08:09.770608   AC0394100   Read-Ack     Max CH water setpoint: 65.00
